### DISCUSSION TITLE
how do you all store the roles?

### DISCUSSION THREAD

**Discussion by laptopmutia** [palkan/action_policy#246] [Q&A] [answered]
currently I have an array field on my model and using `user.roles.includes?("rolename")` to check if they have access or not

**Comment by palkan** [comment, marked as answer]
Usually, I design the authorization model on top of permissions and roles. Permissions are granular and describe business logic restrictions, e.g., "manage_projects", "view_projects", "manage_users", "view_users", etc. Roles act as permission sets and only used to assign typical permissions for users.

Permissions can be stored in an array or JSONB field (I prefer the latter):

```ruby
class User < ApplicationRecord
  attribute :permissions, Permissions::Type.new
end

User.new.permissions #=> Permissions::Set

# Roles only used to synchronize permissions
user.permissions = role.permissions
```

In policies, we check for permissions, not roles:

```ruby
class ApplicationPolicy < ActionPolicy::Base
  # add permissions accessor
  delegate :permissions, to: :user
  # allow everything to adnins by default
  pre_check :allow_admins!

  private

  def allow_admins!
    allow! if admin?
  end
end

class UserPolicy < ApplicationPolicy
  pre_check :ensure_same_account!, only: :show?

  def show?
    permissions.view_users?
  end

  private
  
  def ensure_same_account!
    deny! unless record.account_id == user.account_id
  end
end
```

**Reply by laptopmutia** [reply]
hi its been a while but I got confused about this part
```
class User < ApplicationRecord
  attribute :permissions, Permissions::Type.new
end
```

what is this `Permissions::Type.new` and that line of code meaning?

and for the condition check on the policy models

why your example is just as simple as this `permissions.view_users?`

is there any relations about it and the attribute line code on the `User` class?

is it some kind of hash to object transformation or something?

**Reply by laptopmutia** [reply]
I think its kinda related with this post https://evilmartians.com/chronicles/wrapping-json-based-active-record-attributes-with-classes

**Reply by palkan** [reply]
Yeah, it's a custom active record type. The main reason why we use a custom type is to support updating permissions by updating the object and compact the database representation (so you can have some short identifiers instead of full permission names).

**Comment by laptopmutia** [comment]
any example of controller codes for roles modification that using jsonb? I
also wonder how the strong parameters and the views looks like

**Reply by palkan** [reply]
Here is an example of configuring strong parameters:


```ruby
params.require(:user).permit(:email, :name, permissions:  [Users::Permissions.public_keys])
```

Where `User::Permissions` is a model carrying all the information about possible granular permissions; `public_keys` returned the permissions that can be managed by users (there were also some private permissions, like `admin`).

A client or form submits a hash in a form `{"permissions": { "manage_users": true }}` (all permissions must be specified, not only changed).

**Reply by laptopmutia** [reply]
| (all permissions must be specified, not only changed) 

means all the permission that is returned by `public_keys` right?

**Reply by palkan** [reply]
I meant all permissions you want a user to have, not only the one you want to change
